presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -467,95 +475,272 @@ public RowExpression visitVariableReference(VariableReferenceExpression referenc
+     * For example, (a or b) and (a or c) or ( a or c) returns [[a,b], [a,c], [a,c]]
+     */
+    private List<List<RowExpression>> getSubPredicates(SpecialFormExpression expression)
+    {
+        return extractPredicates(expression.getForm(), expression).stream()
+                .map(LogicalRowExpressions::extractPredicates)
+                .collect(toList());
+    }
+
+    /**
+     * Eliminate a sub predicate if its sub predicates contain its peer.
+     * For example: (a || b) && a = a, (a && b) || b = b
+     */
+    private List<List<RowExpression>> eliminateCommonPredicates(List<List<RowExpression>> predicates)
+    {
+        if (predicates.size() < 2) {
+            return predicates;
+        }
+        // initialize to self
+        int[] reduceTo = IntStream.range(0, predicates.size()).toArray();
+        for (int i = 0; i < predicates.size(); i++) {
+            // Do not eliminate predicates contain non-deterministic value
+            // (a || b) && a should be kept same if a is non-deterministic.
+            // TODO We can eliminate (a || b) && a if a is deterministic even b is not.
+            if (predicates.get(i).stream().allMatch(determinismEvaluator::isDeterministic)) {
+                for (int j = 0; j < predicates.size(); j++) {
+                    if (isSuperSet(predicates.get(reduceTo[i]), predicates.get(j))) {
+                        reduceTo[i] = j; //prefer smaller set
+                    }
+                    else if (isSameSet(predicates.get(reduceTo[i]), predicates.get(j))) {
+                        reduceTo[i] = Integer.min(reduceTo[i], j); //prefer predicates that appears earlier.
+                    }
+                }
+            }
+        }
+
+        return unmodifiableList(stream(reduceTo)
+                .distinct()
+                .boxed()
+                .map(predicates::get)
+                .collect(toList()));
+    }
+
+    /**
+     * Eliminate a sub predicate if its sub predicates contain its peer. Will return null if cannot extract common predicates otherwise return a nested list with flipped form
+     * For example:
+     * (a || b || c || d) && (a || b || e || f)  -> a || b || ((c || d) && (e || f))
+     * (a || b) && (c || d) -> null
+     */
+    private List<List<RowExpression>> extractCommonPredicates(Form rootForm, List<List<RowExpression>> predicates)
+    {
+        if (predicates.isEmpty()) {
+            return null;
+        }
+        Set<RowExpression> commonPredicates = new LinkedHashSet<>(predicates.get(0));
+        predicates.forEach(commonPredicates::retainAll);
+
+        if (commonPredicates.isEmpty()) {
+            return null;
+        }
+        List<RowExpression> remainingPredicates = unmodifiableList(predicates
+                .stream()
+                .map(subPredicates ->
+                        combinePredicates(flip(rootForm),
+                                subPredicates
```

Comment:
Looks like here (line 677 - 679) is implementing `removeAll` with stream. Maybe can replace with a simple function call?

Reply:
`removeAll` is in-place modification, which we probably don't want here.